Elmo Platinum motor controller driver

d-loret opened the pull request:
Initial implementation of the driver for the new Elmo Platinum motor controller drives. The initial implementation focuses on the basic functionality of the drive. All basic modes, cyclic synchronous and profiled, were implemented.

alex-brinkman commented on src/jsd_epd.c:
```diff
@@ -0,0 +1,1305 @@
+ * Public functions
+ ****************************************************/
+
+uint16_t jsd_epd_lc_to_do(char letter_command[2]) {
+  jsd_epd_lc_pair_t  key        = {.lc_chars = letter_command};
+  jsd_epd_lc_pair_t* found_pair = bsearch(
+      &key, jsd_epd_lc_lookup_table,
+      sizeof(jsd_epd_lc_lookup_table) / sizeof(jsd_epd_lc_lookup_table[0]),
+      sizeof(jsd_epd_lc_lookup_table[0]), jsd_epd_compare_lc_keys);
+  return found_pair ? found_pair->do_index : 0x0000;
+}
```
Can you add a few unit tests for this lookup method since it's non-trivial? I wish these were easily computed like in the EGD case :/

d-loret replied:
I have added a unit test querying the currently used letter commands and the special case 0x0000.

alex-brinkman commented on src/jsd_epd.c:
```diff
@@ -0,0 +1,1305 @@
+  MSG_DEBUG("Attempting to map custom EPD PDOs...");
+
+  //////////////// RxPDO Mapping //////////////////////////
+  // TODO(dloret): Not sure if 0x1600-0x1603 are RW in ECAT. EGD used 0x1607 and
+  // 0x1608.
```
If it's working, rm this comment

alex-brinkman commented on src/jsd_epd.c:
```diff
@@ -0,0 +1,1305 @@
+    return 0;
+  }
+
+  // TODO(dloret): not sure if index 0 should be set first to 0, then indexes
+  // 1-8, and then index 0 to the number of mapped objects.
+  // TODO(dloret): Didn't we disable Complete Access somewhere else?
```
If it's working, rm this comment

alex-brinkman commented on src/jsd_epd.c:
```diff
@@ -0,0 +1,1305 @@
+  }
+
+  //////////////// TxPDO Mapping //////////////////////////
+  // TODO(dloret): Not sure if 0x1A00-0x1A03 are RW in ECAT. EGD used 0x1A07 and
+  // 0x1A08.
```
If it's working, rm this comment

alex-brinkman commented on src/jsd_epd.c:
```diff
@@ -0,0 +1,1305 @@
+int jsd_epd_config_LC_params(ecx_contextt* ecx_context, uint16_t slave_id,
+                             jsd_slave_config_t* config) {
+  // TODO(dloret): Verify the types of the corresponding data objects
+  // TODO(dloret): double check that attempting to set 0x0000 (i.e. command not
+  // found) will result in an error. If that is not the case, then the result of
+  // jsd_epd_lc_to_do must be checked before attempting to send the SDO.
```
Please do this before merge, as this will influence the API relating to TL commands!

d-loret replied:
I did test this, so I have removed the comment.

alex-brinkman commented on src/jsd_epd.c:
```diff
@@ -0,0 +1,1305 @@
+  state->pub.drive_temperature = state->txpdo.drive_temperature_deg_c;
+}
+
+void jsd_epd_process_state_machine(jsd_t* self, uint16_t slave_id) {
```
I've gotta trust you on this with hardware-in-loop testing

alex-brinkman reviewed (changes requested):
Great work Daniel! I commend your efforts here.

I like the addition of the elmo common header and your judgement between what should be specific to Gold and Platinum devices. 

My eyes starting glossing over between `epd` and `egd` - do you think it might be worth doing a string replace to make these names more visually discernable?

Finally, can you update the README to include this device in the library?

I did not run these tests since you're the only one with the platinum hardware, assuming this is working well on hardware, feel free to merge once the requested changes are completel and bump the MAJOR release version.

d-loret commented:
Some examples of the tests I did:

![platinum_prof_pos_debug](https://user-images.githubusercontent.com/99947029/224178578-d9e33dff-f463-4199-b552-fd18a1c32322.png)
![platinum_prof_vel_debug](https://user-images.githubusercontent.com/99947029/224178603-c20a2626-027a-43ea-80d2-89fb3a931f02.png)
![platinum_prof_torque_debug](https://user-images.githubusercontent.com/99947029/224178646-9f06902a-9081-46a1-9b0c-76e5dcd6f9b2.png)

alex-brinkman commented:
The command ACK they added to the platinum is a nice feature, it was awkward there was nothing like this in the prof pos command mode on the EGD

d-loret commented:
@alex-brinkman, I have addressed your comments.

About the naming of EGD and EPD, I would prefer if we keep them as they are right now. I personally like how clean the naming of functions, types, and especially enumeration labels end up.

alex-brinkman commented:
Fix that minor merge conflict and then you can merge and mint an updated release. Probably start a new minor release with this sizeable addition.
